File: db4oj/test/src/com/db4o/test/acid/CrashSimulatingBatch.java
/* Copyright (C) 2004 - 2005  db4objects Inc.  http://www.db4o.com */

package com.db4o.test.acid;

import java.io.*;

import com.db4o.foundation.*;
import com.db4o.test.lib.*;


public class CrashSimulatingBatch {
    
    Collection4 writes = new Collection4();
    Collection4 currentWrite = new Collection4();
    
    public void add(byte[] bytes, long offset, int length){
        currentWrite.add(new CrashSimulatingWrite(bytes, offset, length));
    }

    public void sync() {
        writes.add(currentWrite);
        currentWrite = new Collection4();
    }

    public int numSyncs() {
    	return writes.size();
    }
    
    public int writeVersions(String file) throws IOException {
        
        int count = 0;
        int rcount = 0;
        
        String lastFileName = file + "0";
        
        String rightFileName = file + "R" ;
        
        File4.copy(lastFileName, rightFileName);
                
        Iterator4 syncIter = writes.strictIterator();
        while(syncIter.hasNext()){
            
            rcount++;
            
            Collection4 writesBetweenSync = (Collection4)syncIter.next();
            
            if(CrashSimulatingTest.LOG){
                System.out.println("Writing file " + rightFileName + rcount );
            }
            
            RandomAccessFile rightRaf = new RandomAccessFile(rightFileName, "rw");
            Iterator4 singleForwardIter = writesBetweenSync.strictIterator();
            while(singleForwardIter.hasNext()){
                CrashSimulatingWrite csw = (CrashSimulatingWrite)singleForwardIter.next();
                csw.write(rightRaf);
                
                if(CrashSimulatingTest.LOG){
                    System.out.println(csw);
                }
                
            }
            rightRaf.close();
                        
            Iterator4 singleBackwardIter = writesBetweenSync.iterator();
            while(singleBackwardIter.hasNext()){
                count ++;
                CrashSimulatingWrite csw = (CrashSimulatingWrite)singleBackwardIter.next();
                String currentFileName = file + "W" + count;
                File4.copy(lastFileName, currentFileName);
                
                RandomAccessFile raf = new RandomAccessFile(currentFileName, "rw");
                csw.write(raf);
                raf.close();
                lastFileName = currentFileName;
            }
            File4.copy(rightFileName, rightFileName+rcount);
            lastFileName = rightFileName;
            
            
        }
        return count;
    }
    
    
    

}

File: db4on/test/src/com/db4o/test/acid/CrashSimulatingBatch.cs
﻿using System.IO;

namespace com.db4o.test.acid
{
	public class CrashSimulatingBatch
	{
		internal com.db4o.foundation.Collection4 writes = new com.db4o.foundation.Collection4
			();

		internal com.db4o.foundation.Collection4 currentWrite = new com.db4o.foundation.Collection4
			();

		public virtual void Add(byte[] bytes, long offset, int length)
		{
			currentWrite.Add(new com.db4o.test.acid.CrashSimulatingWrite(bytes, offset, length
				));
		}

		public virtual void Sync()
		{
			writes.Add(currentWrite);
			currentWrite = new com.db4o.foundation.Collection4();
		}

		public virtual int NumSyncs()
		{
			return writes.Size();
		}

		public virtual int WriteVersions(string file)
		{
			int count = 0;
			int rcount = 0;
			string lastFileName = file + "0";
			string rightFileName = file + "R";
			CopyFile(lastFileName,rightFileName);
			com.db4o.foundation.Iterator4 syncIter = writes.StrictIterator();
			while (syncIter.HasNext())
			{
				com.db4o.foundation.Collection4 writesBetweenSync = (com.db4o.foundation.Collection4
					)syncIter.Next();
				j4o.io.RandomAccessFile rightRaf = new j4o.io.RandomAccessFile(rightFileName, "rw"
					);
				com.db4o.foundation.Iterator4 singleForwardIter = writesBetweenSync.StrictIterator
					();
				while (singleForwardIter.HasNext())
				{
					com.db4o.test.acid.CrashSimulatingWrite csw = (com.db4o.test.acid.CrashSimulatingWrite
						)singleForwardIter.Next();
					csw.Write(rightRaf);
				}
				rightRaf.Close();
				com.db4o.foundation.Iterator4 singleBackwardIter = writesBetweenSync.Iterator();
				while (singleBackwardIter.HasNext())
				{
					count++;
					com.db4o.test.acid.CrashSimulatingWrite csw = (com.db4o.test.acid.CrashSimulatingWrite
						)singleBackwardIter.Next();
					string currentFileName = file + "W" + count;
					CopyFile(lastFileName,currentFileName);
					j4o.io.RandomAccessFile raf = new j4o.io.RandomAccessFile(currentFileName, "rw");
					csw.Write(raf);
					raf.Close();
					lastFileName = currentFileName;
				}
				rcount++;
				CopyFile(rightFileName,rightFileName + rcount);
				lastFileName = rightFileName;
			}
			return count;
		}
		
		private void CopyFile(string from,string to) 
		{
			File.Delete(to);
			File.Copy(from,to);
		}
	}
}
